File: frontend/src/services/EmployeeService.js
/**
 * This is the "Service" layer in our frontend.
 * It uses 'axios' to send HTTP requests (GET, POST, etc.) to our Spring Boot backend.
 */
import axios from 'axios';

// The URL of our backend API
const EMPLOYEE_API_BASE_URL = "http://localhost:8080/api/v1/employees";

class EmployeeService {

    // Fetch all employees from the backend
    getEmployees() {
        return axios.get(EMPLOYEE_API_BASE_URL);
    }

    // Send a new employee object to the backend to be saved
    createEmployee(employee) {
        return axios.post(EMPLOYEE_API_BASE_URL, employee);
    }

    // Fetch a single employee by their ID
    getEmployeeById(employeeId) {
        return axios.get(EMPLOYEE_API_BASE_URL + '/' + employeeId);
    }

    // Update an existing employee by sending their ID and new data
    updateEmployee(employee, employeeId) {
        return axios.put(EMPLOYEE_API_BASE_URL + '/' + employeeId, employee);
    }

    // Tell the backend to delete an employee by their ID
    deleteEmployee(employeeId) {
        return axios.delete(EMPLOYEE_API_BASE_URL + '/' + employeeId);
    }
}

// We export a 'new' instance of this class so we can use its methods elsewhere.
export default new EmployeeService();

File: backend/src/main/java/com/ems/employee_management_system/EmployeeService.java
package com.ems.employee_management_system;

import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.stereotype.Service;

import java.util.List;

/**
 * This is the "Service" layer.
 * This is where we write our business logic.
 * It sits between the Controller (which handles web requests)
 * and the Repository (which talks to the database).
 */
@Service // Marks this as a Service class.
public class EmployeeService {

    @Autowired // This "injects" our repository automatically. Spring Boot handles it for us.
    private EmployeeRepository employeeRepository;

    // Get all employees from the database
    public List<Employee> getAllEmployees() {
        return employeeRepository.findAll();
    }

    // Save a new employee into the database
    public Employee createEmployee(Employee employee) {
        return employeeRepository.save(employee);
    }

    // Find one employee by their ID
    public Employee getEmployeeById(Long id) {
        return employeeRepository.findById(id).orElseThrow(() -> new RuntimeException("Employee not found with id: " + id));
    }

    // Update an existing employee
    public Employee updateEmployee(Long id, Employee employeeDetails) {
        // First, check if the employee exists
        Employee employee = getEmployeeById(id);
        
        // Update the fields with new information
        employee.setFirstName(employeeDetails.getFirstName());
        employee.setLastName(employeeDetails.getLastName());
        employee.setEmailId(employeeDetails.getEmailId());
        employee.setSalary(employeeDetails.getSalary());
        employee.setDob(employeeDetails.getDob());
        employee.setLocation(employeeDetails.getLocation());
        employee.setMobileNumber(employeeDetails.getMobileNumber());
        
        // Save the updated employee back to the database
        return employeeRepository.save(employee);
    }

    // Delete an employee by their ID
    public void deleteEmployee(Long id) {
        Employee employee = getEmployeeById(id);
        employeeRepository.delete(employee);
    }
}
